### PR TITLE
tth jet uncertainty, b-tag weighting

### PR DESCRIPTION
**Pull request by jpata**
Forward port of vhbb/cmssw#68.

### CODE REVIEW

**Comment by cbernet** [comment]
@gpetruc  @mariadalfonso
could you please just confirm that you are ok with this change for CMG?

File: PhysicsTools/Heppy/python/analyzers/objects/JetAnalyzer.py [hunk 1]
```diff
@@ -100,6 +101,12 @@ def process(self, event):
         if self.doJEC:
 #            print "\nCalibrating jets %s for lumi %d, event %d" % (self.cfg_ana.jetCol, event.lumi, event.eventId)
             self.jetReCalibrator.correctAll(allJets, rho, delta=self.shiftJEC, metShift=self.deltaMetFromJEC)
+
+        for delta, shift in [(1.0, "JECUp"), (0.0, ""), (-1.0, "JECDown")]:
```
Comment (arizzi) [inline PhysicsTools/Heppy/python/analyzers/objects/JetAnalyzer.py:105]: aren't we computing twice the 0.0  ?
shouldn't we make configurable the lisft of shifts and perhaps get rid of L103? we can perhaps use the first shift in the list as new default and store all the others as "corr"+shift ?
@gpetruc what do you think?

Comment (jpata) [inline PhysicsTools/Heppy/python/analyzers/objects/JetAnalyzer.py:105]: yes, this isn't the most optimal approach. I wanted it to be independent of doJEC. we could put the delta loop inside doJEC and use the first one as the default in this case.

**Comment by cbernet** [comment]
@arizzi @gpetruc 
Hi guys, sorry but I'm still not sure whether you want me to merge this. 
Could you please give me your green light? 
C

**Comment by arizzi** [comment]
I'm ok with it but I let @gpetruc comment
perhaps we should make it more configurable? (not a showstopper anyhow)

**Comment by cbernet** [comment]
Yes, not a showstopper, I’m willing to merge now, but I want to be sure CMG is unaffected. 
ccing maria as well.

**Comment by mariadalfonso** [comment]
Indeed I want to test the performances/changes/improvements of this PR. 
However in the CMG we are still using the heppy_7_4_3  while this PR is on the heppy_7_4_5.
Can Joosep back in the heppy_7_4_3 as well ?

Maria

**Comment by jpata** [comment]
@mariadalfonso Here is the original 743 PR: https://github.com/vhbb/cmssw/pull/68
This is a forward port to 745 as requested by @cbernet .

**Comment by arizzi** [comment]
maria this just adds stuff..

**Comment by cbernet** [comment]
Hi think it’s a bit painful to ask Joosep to backport.
Actually I asked him to port these developments forward… :-) 

So here is what I propose: 
- I merge now. 
- When CMG gets this development, they test it and let us know.
- If needed, we can always revert the commit

Ok Maria? 

C

**Comment by mariadalfonso** [comment]
Hi Colin, ok let's proceed how you suggested. 
We will check when CMG move forward.
Maria

**Comment by arizzi** [comment]
@jpata this seem to have trouble when recalibrateJet is not set (the recalibrator class is not initialized), so we can do that only if we have correctly configured a calibrator... in addition if we do not  recalibrate, can we fetch the information from the pat::Jet or is just not there?
in any case the fact that we call twice the nominal calibration should be fixed. Can you please  take care of fixing these issues and make a new PR with the fixes?

**Comment by jpata** [comment]
OK I'm taking a look now.

**Comment by jpata** [comment]
Please check PR https://github.com/cbernet/cmssw/pull/44. I didn't understand what you meant by:
"in addition if we do not recalibrate, can we fetch the information from the pat::Jet or is just not there?" IIRC we save the corr through

```
getattr(jet, "corr", -99)
```

**Comment by arizzi** [comment]
I mean if we do not want to recalibrate in Heppy, and stay with the energy from miniaod, is there a way to access JecUp/Down from pat::Jet or they  are just not stored?

**Comment by gpetruc** [comment]
On Thu, Jul 16, 2015 at 12:00 PM, arizzi notifications@github.com wrote:

> I mean if we do not want to recalibrate in Heppy, and stay with the energy
> from miniaod, is there a way to access JecUp/Down from pat::Jet or they are
> just not stored?

they are not computed nor stored, so no

**Pull request merged**